### ISSUE TITLE
Implement wonky walls like in DK2

### ISSUE THREAD

**Issue by ufdada** [tonihele/OpenKeeper#108]
It's just a visual thing, but this would definitely make it look closer to the original.

This was actually achieved with **Incremental Delaunay Triangulation**.

Readings:
* https://www.google.de/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0CCYQFjAAahUKEwjf2KahpITJAhVGEiwKHWZYBqU&url=http%3A%2F%2Fwww.karlchenofhell.org%2Fcppswp%2Flischinski.pdf&usg=AFQjCNERvoyhQYMwgtVXdNJYlZukkk8Kiw&sig2=E83lExb014YunXw9nbgkpg
* http://www.ti.inf.ethz.ch/ew/courses/CG13/lecture/Chapter%207.pdf
* https://www.mi.fu-berlin.de/en/math/groups/discgeom/dates/2nd_ERC_Workshop/slides/inc3-talk.pdf

**Comment by tonihele** [comment]
Terrain has wibble & lean values. Both vertical & horizontal which probably contribute on how much wonkyness is applied. Stone doesn't have much, but regular dirt is all over the place.

**Comment by belgianguy** [comment]
I used to think that this article was referenced somewhere on this repository, but I can't find it,  @(tonihele) also commented on it. As did Robin Green (Ex-Bullfrog). Knowing when to split or flip.

https://simonschreibt.de/gat/dungeon-keeper-2-walls/

**Comment by tonihele** [comment]
Here is also a little something. Shader based wonkyness:
https://keeperklan.com/threads/7062-Warped-walls-recreation?p=55292#post55292

**Comment by Trass3r** [comment]
I also think they just used some simple vertex manipulations to create the effect.
That still seems tricky to do in OpenKeeper though seeing how complicated the terrain rendering is. It's all done in many small chunks according to RenderDoc.

**Comment by tonihele** [comment]
Physics-wise I'm also routing for non shader based solution...

**Comment by rainlizard** [comment]
> Physics-wise I'm also routing for non shader based solution...

Not sure what you mean by physics, but just a reminder that they used perfect cube collision boxes. (collisions do not match the warping effect)

Unless your desire is to be even more accurate and _better_ than DK2.

**Comment by tonihele** [comment]
Good catch. Didn't actually know that, but it makes sense. DK 2 doesn't really have physics in a sense. Everything is quite simple, such as shadows. But we were planning a bit to have Bullet physics. It will also make things maybe easier to implement even. If it totally breaks up the game, then not. But worth trying.

**Comment by Trass3r** [comment]
Played around a bit with this warping effect:
https://playground.babylonjs.com/#QMFPPE#250

**Comment by ArchDemons** [comment]
I think we should use Perlin noise (or another pseudorandom generator) because in the same vertices of adjacent block must have equal coordinatase

**Comment by Trass3r** [comment]
That's easy. The problem is the vertex displacement will affect the normals and tangents.

**Comment by RobertZenz** [comment]
I just got reminded that I did [implement something similar for a bloxel game engine](https://gitlab.com/quadracoatl/quadracoatl/-/blob/master/jme-client/src/org/quadracoatl/jmeclient/meshers/DisplacingSimpleMesher.java) written with jME. Though, I did it when creating the mesh, it might be getter to dispklace the grid in the shader.

**Comment by Trass3r** [comment]
PoC with a slight offset: https://github.com/Trass3r/OpenKeeper/commits/wobblywalls
![image](https://user-images.githubusercontent.com/404623/180466045-d47cdf9c-504a-497c-bfd8-1789803fd390.png)

**Comment by tonihele** [comment]
Already makes a huge difference, doesn't look so dul

**Comment by Trass3r** [comment]
Yeah indeed.
But it would better  if you could extend the default materials/shaders somehow instead of all this copy paste, and you need a way to only assign the shader to map geometry and the selector wireframe (or a material parameter). For some reason it even affects the water and torches even though it shouldn't.

**Comment by tonihele** [comment]
The default shaders in jME are just that, they are very generic and support all features we don't really use. I would imagine that in the future we would just have bunch of custom light weight shaders for everything to provide that cool Dungeon Keeper 2 feeling. Hopefully with a modern twist of course.

**Comment by Trass3r** [comment]
Yeah I just thought the same, sooner or later you'll have fully custom shaders anyway.

**Comment by Trass3r** [comment]
Updated thoughts on a shader-based solution vs computing the map mesh on the CPU:
- can use instanced cubes to reduce the amount of geometry data which is good esp. since the sides need to be subdivided once at least
- but you'll render quite some faces that are invisible
- requires a special shader with knowledge about the world setup baked in
- requires uploading the map as a texture to drive the distortion directions
- lots of vertex work every frame, even though the geometry is relatively static
- effort gets multiplied by Z pre pass or shadow rendering
- similarly, pixel shader work to reconstruct the normals etc
- the distorted vertices only exist on the GPU, so shadows work but not stuff computed on the CPU like collisions maybe

**Comment by Trass3r** [comment]
For a non-shader solution `AssetUtils.translateToTile` might be a good starting point as it's called by `handleTop`/`handleSide` and the room constructors.
But you can't just use the position buffer there, the rotation of the quads also needs to be taken into account.

**Comment by belgianguy** [comment]
<img width="3840" height="1080" alt="Image" src="https://github.com/user-attachments/assets/4cf774e0-d3d8-41e0-9ff0-37e4d42c40c6" />
<img width="3840" height="1080" alt="Image" src="https://github.com/user-attachments/assets/95f8907e-6a0e-4402-920b-9dcebad04f65" />
<img width="3840" height="1080" alt="Image" src="https://github.com/user-attachments/assets/e367fa16-e91c-488e-8298-b0e3346425f5" />

An experimental implementation on a fresh `feature/wobbly-walls` branch: https://github.com/belgianguy/OpenKeeper/tree/feature/wobbly-walls

The implementation uses a deterministic shared world-grid control cage:

- Each integer terrain corner receives a stable 3D offset.
- Every tile sharing that corner uses exactly the same offset, keeping different terrain types watertight.
- Existing DKII top and wall meshes are deformed against the four surrounding cage corners.
- Wall deformation fades to zero at floor height.
- Normals and tangents are regenerated after deformation.
- The selection outline uses the same displaced corners.
- Deformation happens once before batching, rather than as post-processing or a per-frame shader.

Credit and references:

- [Simon’s Dungeon Keeper 2 walls article](https://simonschreibt.de/gat/dungeon-keeper-2-walls/) provided the main visual analysis and showed that the original source meshes are straight but appear deformed in-game.
- [Trass3r’s earlier OpenKeeper `wobblywalls` experiment](https://github.com/Trass3r/OpenKeeper/commits/wobblywalls) demonstrated a world-position hash implemented in a vertex shader. We investigated this approach but ultimately used CPU-side deformation so geometry, normals, tangents and outlines remain synchronized.
- [Quadracoatl’s `DisplacingSimpleMesher`](https://gitlab.com/quadracoatl/quadracoatl/-/blob/master/jme-client/src/org/quadracoatl/jmeclient/meshers/DisplacingSimpleMesher.java) inspired the shared-lattice approach: neighboring faces derive their displaced vertices from the same world coordinates.
- The linked Babylon.js experiments ([revision 250](https://playground.babylonjs.com/#QMFPPE#250) and [revision 373](https://playground.babylonjs.com/#QMFPPE#373)) were useful practical and visual references.
- Thanks to @ufdada for opening this issue and collecting the original Delaunay references.

This is not a port of any of those implementations and does not currently implement Delaunay triangulation. The original Bullfrog developer discussion connects incremental Delaunay triangulation primarily with the dynamically changing map/navigation topology. Our current visual deformation is a custom shared-cage implementation inspired by the references above.

OpenKeeper already parses DKII’s per-terrain `wibbleH`, `wibbleV`, and lean values, but the experiment currently uses visually tuned constants. Integrating those values without reopening seams between terrain types is a possible follow-up.

**Comment by ufdada** [comment]
I just messed around with AI and let it create a 2D demo with digging and pathfinding. you can also adjust the wall deformation.

[dig-delaunay-demo-js.html](https://github.com/user-attachments/files/31185000/dig-delaunay-demo-js.html)
